### PR TITLE
adding backward capability for potrf (Cholesky)

### PR DESCRIPTION
**Pull request by jmxpearson**
- cf #440 
- Implementation based on Iain Murray's note (arXiv 1602.07527)
- Uses the symbolic method, which is easier to code and competitive for
  small arrays (according to experiments in the supplement to that paper)
- Unfortunately, the framework in `test_autograd` fails for the `Function` subclass because the outputs of `Potrf` are not independent, so the Jacobian calculation isn't correct. I implemented a different set of tests based on Murray's [here](https://github.com/imurray/chol-rev/blob/master/python/chol_rev_demo.py)

### CODE REVIEW

File: torch/autograd/_functions/linalg.py [hunk 1]
```diff
@@ -102,3 +102,44 @@ def backward(ctx, grad_output, grad_LU=None):
...
+        return fact
+
+    @staticmethod
+    def Phi(A):
+        """
+        Return lower triangle of A and halve the diagonal.
+        """
+        B = A.tril()
+        for i in range(B.size(0)):
+            B[i, i] = B[i, i] * 0.5
```
Comment (fmassa) [inline torch/autograd/_functions/linalg.py]: Might be more efficient to replace this loop with something like
```python
# need to make sure that mat is in the same device as A
mat = torch.ones(A.size()) * (-0.5) + 1
mat = Variable(mat)
B = B * mat
```

File: test/test_potrf.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import torch
+import numpy as np
+from test_autograd import _make_cov
+from torch.autograd import Variable
+from common import TestCase, run_tests
+
+from torch.autograd._functions.linalg import Potrf
+
+class TestPotrf(TestCase):
```
Comment (fmassa) [inline test/test_potrf.py]: Can you skip this test if lapack is not found on the system?

File: torch/autograd/_functions/linalg.py [hunk 3]
```diff
@@ -102,3 +102,44 @@ def backward(ctx, grad_output, grad_LU=None):
...
+
+    @staticmethod
+    def forward(ctx, a, upper=True):
+        ctx.upper = upper
+        fact = torch.potrf(a, upper)
+        ctx.save_for_backward(fact)
+        return fact
+
+    @staticmethod
+    def Phi(A):
```
Comment (fmassa) [inline torch/autograd/_functions/linalg.py]: can you make this method lowercase?

**Review by fmassa — CHANGES_REQUESTED** [review]
Did a first pass on the code, overall looks good but I have a few comments. I didn't check the correctness of the gradients, will wait for jenkins on `test_autograd`.
Thanks for the PR!

File: torch/autograd/_functions/linalg.py (hunk 1 above)
Comment (jmxpearson) [inline torch/autograd/_functions/linalg.py]: Makes sense. I didn't have a good intuition for whether the speed or the allocation was the bigger issue.

Comment (jmxpearson) [inline torch/autograd/_functions/linalg.py]: What's the idiom for making sure `mat` is on the same device? Something like
```python
mat = A.data.new(torch.eye(A.size(0))) * (-0.5) + 1
```
?

**Comment by jmxpearson** [comment]
I also fixed the `flake8` warnings that were causing Travis to bail. Have made the changes requested and will push again when we get the bit about devices resolved.

File: torch/autograd/_functions/linalg.py (hunk 1 above)
Comment (fmassa) [inline torch/autograd/_functions/linalg.py]: So, it turns out that there is not yet a version of `torch.eye` that works directly on the GPU.
I implemented it in https://github.com/pytorch/pytorch/pull/2395 , so once it's merged, you could do something like
```python
mat = torch.eye(A.size(0), out=A.data.new()) * (-0.5) + 1
```

Comment (jmxpearson) [inline torch/autograd/_functions/linalg.py]: I'll wait for that to merge and then push.

Comment (pavanky) [inline torch/autograd/_functions/linalg.py]: It would probably be easier to do the following ?
```
B = B - 0.5 * torch.diag(torch.diag(B))
```

Comment (fmassa) [inline torch/autograd/_functions/linalg.py]: Definitely easier, and just slightly less memory efficient. I'd go for @pavanky solution!

**Comment by jmxpearson** [comment]
Good deal.

File: torch/autograd/_functions/linalg.py [hunk 4]
```diff
@@ -102,3 +102,44 @@ def backward(ctx, grad_output, grad_LU=None):
...
+            L = L.t()
+            grad_output = grad_output.t()
+
+        # make sure not to double-count variation, since
+        # only half of output matrix is unique
+        Lbar = grad_output.tril()
+
+        P = Potrf.Phi(torch.mm(L.t(), Lbar))
+        S = torch.gesv(P + P.t(), L.t())[0]
+        S = torch.gesv(S.t(), L.t())[0]
```
Comment (pavanky) [inline torch/autograd/_functions/linalg.py]: Wouldn't it be more efficient to use btriact on `L.t()` and call `btrisolve` twice instead of gesv. ?
http://pytorch.org/docs/master/torch.html#torch.btrisolve

Comment (fmassa) [inline torch/autograd/_functions/linalg.py]: Maybe, but `btrifact` and `btrisolve` are not yet implemented for `Variable`s, so it would require the function to be implemented via `@once_differentiable`, limiting it's use to first-order derivatives.

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
thanks a lot John!
